### PR TITLE
10015 data export task

### PR DESCRIPTION
**Pull request by DevneyHamilton**

### CODE REVIEW

File: app/models/loan.rb [hunk 1]
```diff
@@ -211,4 +211,24 @@ def healthy?
...
+  end
+
+  def sum_of_repayments(start_date: nil, end_date: nil)
+    return nil if transactions.empty?
+    transactions.by_type("repayment").in_date_range(start_date, end_date).map { |t| t.amount }.sum
+  end
+
+  def change_in_interest(start_date: nil, end_date: nil)
+    return nil if transactions.empty?
+    transactions.in_date_range(start_date, end_date).map { |t| (t.change_in_interest || 0) }.sum
```
Comment (DevneyHamilton) [inline app/models/loan.rb]: With manual testing, I found there are transactions with null change_in_interest (which err'ed on `sum`). I need to figure out why t.change_in_interest can be null sometimes. Either it should not be null or if there are transactions with null change_in_interest, should the this method return null?

File: app/models/loan.rb [hunk 2]
```diff
@@ -211,4 +211,24 @@ def healthy?
...
+  end
+
+  def change_in_principal(start_date: nil, end_date: nil)
+    return nil if transactions.empty?
+    transactions.in_date_range(start_date, end_date).map { |t| (t.change_in_principal || 0) }.sum
```
Comment (DevneyHamilton) [inline app/models/loan.rb]: same issue as change_in_interest

File: app/models/standard_loan_data_export.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  # Subclass exists only to implement process_data. No additional public methods should be added to this subclass.
+  def process_data
+    @child_errors = []
+    data = []
+    data << header_row
+    Loan.find_each do |l|
+      begin
+        data << hash_to_row(loan_data_as_hash(l))
+      rescue => e
+        @child_errors << {loan_id: l.id, message: e.message}
```
Comment (DevneyHamilton) [inline app/models/standard_loan_data_export.rb:37]: This uses the same format as the errors in the loan update task. There isn't any enforcement of this format except that the task/show view assumes it.

**Comment by DevneyHamilton** [comment]
I still need to look into what happens when qb is not connected and other cases we should handle.

File: app/controllers/admin/data_exports_controller.rb [hunk 4]
```diff
@@ -5,6 +5,22 @@ def new
     @data_export = @export_class.new
   end
 
+  def create
+    # TODO: update to take form params (issue 10017); adding here to create taskable association
+    @data_export = StandardLoanDataExport.create(
+      division_id: division.id,
+      locale_code: "en",
```
Comment (DevneyHamilton) [inline app/controllers/admin/data_exports_controller.rb:12]: generalized in 10016

File: app/controllers/admin/data_exports_controller.rb [hunk 5]
```diff
@@ -5,6 +5,22 @@ def new
...
+      start_date: Date.parse("2019-01-01"),
+      end_date: Date.parse("2019-09-01")
+    )
+    Task.create(
+      job_class: DataExportTaskJob,
+      job_type_value: 'data_export_task_job',
+      activity_message_value: 'task_enqueued',
+      taskable: @data_export
```
Comment (DevneyHamilton) [inline app/controllers/admin/data_exports_controller.rb:20]: it's feasible to set the taskable association elsewhere if that would be better?

File: app/models/accounting/quickbooks/accounts_not_selected_error.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+module Accounting
+  module Quickbooks
+    class AccountsNotSelectedError < StandardError
```
Comment (DevneyHamilton) [inline app/models/accounting/quickbooks/accounts_not_selected_error.rb:3]: When testing to make sure that the StandardLoanDataExport is not affected by the qb connection or accounts settings, I realized this error should be raised for the UpdateAllLoanJob, so I slipped this in here.

File: app/models/accounting/transaction.rb [hunk 7]
```diff
@@ -80,10 +80,22 @@ class Accounting::Transaction < ApplicationRecord
 
   scope :standard_order, -> {
     joins("LEFT OUTER JOIN options ON options.option_set_id = #{loan_transaction_type_option_set.id}
-      AND options.value = accounting_transactions.loan_transaction_type_value").
-    order(:txn_date, "options.position", :created_at)
+      AND options.value = accounting_transactions.loan_transaction_type_value")
+      .order(:txn_date, "options.position", :created_at)
   }
   scope :interest_type, -> { where(loan_transaction_type_value: LOAN_INTEREST_TYPE) }
```
Comment (DevneyHamilton) [inline app/models/accounting/transaction.rb]: style issues here will be fixed when 10013 and 10014 are merged and this branch is rebased

File: app/models/accounting/transaction_data_missing_error.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+module Accounting
+  class TransactionDataMissingError < StandardError
+    def message
+      "Loan has transactions with no change in interest. Do you need to run a QB update?"
```
Comment (DevneyHamilton) [inline app/models/accounting/transaction_data_missing_error.rb]: Need to translate this

File: app/models/loan.rb [hunk 9]
```diff
@@ -211,4 +215,34 @@ def healthy?
...
+  def sum_of_repayments(start_date: nil, end_date: nil)
+    return nil if transactions.empty?
+    transactions.by_type("repayment").in_date_range(start_date, end_date).map { |t| t.amount }.sum
+  end
+
+  def change_in_interest(start_date: nil, end_date: nil)
+    return nil if transactions.empty?
+    changes = transactions.in_date_range(start_date, end_date).map { |t| (t.change_in_interest) }
+    if changes.any? { |i| i.nil? }
+      raise Accounting::TransactionDataMissingError
```
Comment (DevneyHamilton) [inline app/models/loan.rb]: Without a custom error, what the user sees is many loans with `"NoMethodError: no method '+' for nil.` This is just an attempt to make it clear that that error should be fixed by updating all loans. Happy for suggestions on better ways to do this.

**Comment by DevneyHamilton** [comment]
Failing on CI with ```0.21s$ bundle exec rspec
bundler: command not found: rspec
Install missing gem executables with `bundle install`
The command "bundle exec rspec" exited with 127.
cache.2
store build cache``` but 10016 is not. I am not sure what to do about this?

File: app/models/loan.rb [hunk 10]
```diff
@@ -69,6 +69,10 @@ class Loan < Project
   scope :related_loans, ->(loan) { loan.organization.loans.where.not(id: loan.id) }
 
   delegate :name, :country, :postal_code, to: :organization, prefix: :coop
+  delegate :name, to: :primary_agent, prefix: true, allow_nil: true
+  delegate :name, to: :secondary_agent, prefix: true, allow_nil: true
+  delegate :text, to: :status, prefix: true, allow_nil: true
```
Comment (meltheadorable) [inline app/models/loan.rb]: discussed in pair

File: app/models/loan.rb (hunk 9 above)
Comment (meltheadorable) [reply app/models/loan.rb]: `raise Account::TransactionDataMissingError if changes.any(&:blank?)` would shorten this up a bit

**Review by meltheadorable — APPROVED** [review]
Looks good, beyond the changes we discussed in pairing.

**Comment by emitche** [comment]
@DevneyHamilton Does this still need revision?

**Comment by DevneyHamilton** [comment]
> @DevneyHamilton Does this still need revision?

@emitche No, it doesn't - sorry. I marked it 'complete' because this is 'behind the scenes' of 10016/10017, so I think they can be qa'ed together. The actual report contents can be qa'ed via the download attachment file link in 10018.

**Comment by emitche** [comment]
Tests ran successfully on my local.

**Pull request merged**